constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/Windows.java

```diff
@@ -1,267 +1,359 @@
+    this.segments = new IntervalMap<>(interval, value);
   }
 
-  public void add(final Window window) {
-    this.windows.add(window);
+  public Windows(final IntervalMap<Boolean> segments) {
+    this.segments = segments;
   }
 
-  public void addAll(final Windows other) {
-    this.windows.addAll(other.windows);
+  public Windows(final Boolean bool) {
+    this.segments = new IntervalMap<>(bool);
   }
 
-  public void addPoint(final long quantity, final Duration unit) {
-    this.add(Window.at(quantity, unit));
+  public Windows and(final Windows other) {
+    return new Windows(IntervalMap.map2(
+        this.segments, other.segments,
+        (l, r) -> {
+          if (l.isPresent() && r.isPresent()) {
+            return Optional.of(l.get() && r.get());
+          } else if (l.isPresent()) {
+            return l.get() ? Optional.empty() : Optional.of(Boolean.FALSE);
+          } else if (r.isPresent()) {
+            return r.get() ? Optional.empty() : Optional.of(Boolean.FALSE);
+          } else {
+            return Optional.empty();
+          }
+        }
+    ));
   }
 
-  public static Windows union(final Windows left, final Windows right) {
-    final var result = new Windows(left);
-    result.addAll(right);
-    return result;
+  public Windows or(final Windows other) {
+    return new Windows(IntervalMap.map2(
+        this.segments, other.segments,
+        (l, r) -> {
+          if (l.isPresent() && r.isPresent()) {
+            return Optional.of(l.get() || r.get());
+          } else if (l.isPresent()) {
+            return l.get() ? Optional.of(true) : Optional.empty();
+          } else if (r.isPresent()) {
+            return r.get() ? Optional.of(true) : Optional.empty();
+          } else {
+            return Optional.empty();
+          }
+        }
+    ));
   }
 
-
-  public void subtract(final Window window) {
-    this.windows.subtract(window);
+  public Windows add(final Windows other) {
+    return new Windows(IntervalMap.map2(
+        this.segments, other.segments,
+        (l, r) -> {
+          if (l.isPresent() && r.isPresent()) {
+            return Optional.of(l.get() || r.get());
+          } else if (l.isPresent()) {
+            return l;
+          } else return r;
+        }
+    ));
   }
 
-  public void subtractAll(final Windows other) {
-    this.windows.subtractAll(other.windows);
+  public Windows not() {
+    //should not be a subtraction because then if it was null originally, then subtracting original from forever
+    //  yields true where once was null, which isn't good. we want a simple inversion of true and false here, without
+    //  filling nulls.
+    return new Windows(this.segments.map($ -> $.map(b -> !b)));
   }
 
-  public void subtract(final long start, final long end, final Duration unit) {
-    this.subtract(Window.between(start, end, unit));
+  public Optional<Pair<Duration, Interval.Inclusivity>> minTrueTimePoint(){
```

Comment:
Yes currently it will return `Duration.MIN_VALUE`. And I agree this would be difficult to maintain with unbounded intervals, instead of infinite durations.

I hate to say it, but I think allowing unbounded intervals is unlikely to happen any time soon though, if ever, since I think its pretty low priority. So I'd argue that if we ever get around to it, we deal with `minTrueTimePoint` when it becomes a blocker.